### ISSUE TITLE
Dragon Age Inquisition fails to start sporadically

### ISSUE THREAD

**Issue by Alexander--** [doitsujin/dxvk#293]
Running the game with dxvk 0.42 fails 90% of time, but succeeds in remaining 10%. The game shows company logo and even renders into animation (I was able to get to the "press any button" prompt). The exact conditions, that allow the game to start, are unclear.

The results were obtained with radv/mesa-18.0 and wine-3.5 on Arch Linux. 

Log output:

````
info:  Enabled instance layers:
info:  Enabled instance extensions:
info:    VK_KHR_surface
info:    VK_KHR_win32_surface
WARNING: radv is not a conformant vulkan implementation, testing use only.
info:  AMD RADV POLARIS11 (LLVM 6.0.0):
info:    Driver: 18.0.0
info:    Vulkan: 1.0.57
ERROR: Could Not Get Primary Adapter Handle <- this line does not always appear??
info:  D3D11CreateDevice: Probing D3D_FEATURE_LEVEL_11_1
err:   D3D11CreateDevice: Requested feature level not supported
info:  D3D11CreateDevice: Probing D3D_FEATURE_LEVEL_11_0
info:  D3D11CreateDevice: Using feature level D3D_FEATURE_LEVEL_11_0
warn:  Vulkan extension VK_KHR_maintenance2 not supported
info:  Enabled device extensions:
info:    VK_KHR_descriptor_update_template
info:    VK_KHR_sampler_mirror_clamp_to_edge
info:    VK_KHR_maintenance1
info:    VK_KHR_shader_draw_parameters
info:    VK_KHR_swapchain
warn:  DXGI: MakeWindowAssociation: Ignoring flags
warn:  D3D11Device::GetDeviceRemovedReason: Stub
````

**Comment by doitsujin** [comment]
Hm, interesting.

I own the game so in theory I could test this myself, but please tell me one thing, how the hell did you get Origin to run? All the various instructions I've found fail for one reason or another.

**Comment by portentum** [comment]
Sergey Pavlyuchenko's [AppDB](https://appdb.winehq.org/objectManager.php?sClass=version&iId=26175&iTestingId=101505) instructions work fine for installing Origin. Step 9 is important as Origin likes to screw up its file permissions. 
I just re-tested the install steps and Origin crashed on first launch after updating, but runs fine on subsequent launches.
I also use this [patch](https://bugs.winehq.org/attachment.cgi?id=59646&action=diff&context=patch&collapsed=&headers=1&format=raw) from [bug#43187](https://bugs.winehq.org/show_bug.cgi?id=43187#c2) to prevent OriginWebHelperService.exe from crashing, but I don't think it was required. The crash just annoyed me.

That said, I have had zero success getting Dragon Age Inquisition to run either. It always just crashes on launch, taking winedbg with it, giving me no backtrace. It crashes slightly later, showing the pulsing Inquisition loading logo at the bottom right of a black screen for a couple of seconds, with DXVK. With wined3d it just crashes immediately on launch without displaying anything at all.

[DragonAgeInquisition_d3d11.log](https://gist.github.com/portentum/388f4c43552f63a4412f4a40812e5a8a)
[DragonAgeInquisition_dxgi.log](https://gist.github.com/portentum/fd2557cb7b57eda89d61dc51a894c7c8)

**Comment by doitsujin** [comment]
@portentum well they don't work for me, which is why I'm asking in the first place.

Anyway, if the game just crashes for no apparent reason there's not much I can do without a backtrace.

**Comment by Alexander--** [comment]
@doitsujin @portentum 

> That said, I have had zero success getting Dragon Age Inquisition to run either. It always just crashes on launch, taking winedbg with it, giving me no backtrace. With dxvk Inquisition crashes slightly later, after showing the pulsing Inquisition loading logo at the bottom right of a black screen for a couple of seconds.

This is what usually happens for me as well. But there is a small chance, that it runs further: I have seen it reach the marching soldiers animation *twice* out of dozens attempts. Never managed to get a stack trace either.

**Comment by steemandlinux** [comment]
https://bugs.winehq.org/show_bug.cgi?id=39229
https://bugs.winehq.org/show_bug.cgi?id=40942

Denuvo protection

**Comment by doitsujin** [comment]
@steemandlinux Denuvo is not *necessarily* the issue here. I'm not sure why people still think that Denuvo doesn't work with wine, plenty of Denuvo-protected games have been running for quite a while now.

**Comment by portentum** [comment]
@Alexander-- @steemandlinux @doitsujin 
[It works.](https://imgur.com/0JyutFS)
While other games don't care, DA:I can't be on a partition mounted noexec it would seem. Changing that mount flag allowed the game to launch to the marching soldiers, click to continue screen.

After that it would freeze when clicking to continue to the menu, seemingly caused by trying to go fullscreen. Editing ProfileOptions_profile to disable fullscreen allowed me to start a new game, create a character, and run around a bit.

**Comment by doitsujin** [comment]
@portentum Thanks for letting us know. Can you test if the fullscreen mode works correctly with bb3e3c9e4f15e5232d1a60ffbda00d9fd55064c4?

**Comment by portentum** [comment]
@doitsujin 
Tested with adb0f5e16b85623cd5b4409294f7df1007ac2209, the game still freezes (no crash, application just hangs indefinitely chewing up two CPU cores) when attempting to fullscreen.


[DragonAgeInquisition.orig_d3d11.log](https://github.com/doitsujin/dxvk/files/1924789/DragonAgeInquisition.orig_d3d11.log)
[DragonAgeInquisition.orig_dxgi.log](https://github.com/doitsujin/dxvk/files/1924790/DragonAgeInquisition.orig_dxgi.log)

**Comment by doitsujin** [comment]
The logs aren't particularly helpful, I don't know what the game is trying to do and what it is expecting. Can you try to record an apitrace on Windows?

There's basically no reason why switching to fullscreen mode shouldn't work at this point, unless the game relies on undocumented implementation details (which is entirely possible, my latest commits fixed a couple of those issues).

**Comment by Alexander--** [comment]
@doitsujin 

I don't know if this may be helpful in some way, but here are some unnecessary details about my setup: the couple of times when I was able to successfully launch the game, it's window was forced into a fixed size by the tiling windows manager (i3). The WM denied window size change to fullscreen, and the game was able to run fine (so it does not really care if the size change succeeds, just that size is usable for engine).

But there appears to be some kind of race — if the size change is correctly timed, it works out, otherwise it crashes.

**Comment by doitsujin** [comment]
Yeah, sounds like a DXGI issue. Sadly those are next to impossible to debug without getting the application to run, and Origin *still* refuses to work on my setup.

**Comment by portentum** [comment]
I'm having trouble getting an apitrace out of DAI on Windows. Origin gets in the way, and my little attempt to use a bat file renamed to exe just resulted in an error about 16-bit exes not being supported. I don't know enough about this stuff on Windows to know what else to try off the top of my head.

I'll pick it back up later tonight and continue trying to get an apitrace.

**Comment by portentum** [comment]
@doitsujin 
A bat-to-exe utility came to my rescue. I was able to get an apitrace more sooner than I'd expected.
[DAI.win7.trace](https://ufile.io/vwhan) ([alternate download](https://mega.nz/#!WK4T0DzS!ECHNh6Ya_F50XYkBJRJ6suzgW-lyMiiM4s3V1yLxhyk))

**Comment by Alexander--** [comment]
@doitsujin I noticed, that 2f20a61342 had botched "Closes " message. This issue can be closed now, right?

**Comment by doitsujin** [comment]
@Alexander-- That should have referred to #314, this issue is unrelated.

**Comment by doitsujin** [comment]
I finally got Origin and the game to work on my setup and can reproduce the fullscreen freeze. While I know what *causes* it - namely, Thread A waiting for a callback on Thread B to finish while Thread B blocks on a mutex that is currently held by Thread A - I don't know how to fix this properly. If anyone knows something about thread safety guarantees of IDXGISwapChain, please educate me.

**Comment by Nicholas-Steel** [comment]
The game seems to be working fine here within Windows 10 with the current release of DXVK, the only thing I've noticed is that it forces the game in to Fullscreen Exclusive Mode even with the game set to Windowed Fullscreen (this isn't the only game where I've noticed DXVK doing this).

**Comment by doitsujin** [comment]
Not sure why you're digging up a 6-year old issue but some Vulkan drivers on Windows will treat every fullscreen window as fullscreen exclusive. It's not a DXVK-exclusive quirk.

**Comment by Nicholas-Steel** [comment]
You're right, this wasn't the best way to go about this. Sorry about that.
